### ISSUE TITLE
Consistent Use of Context

### ISSUE THREAD

**Issue by devigned** [kubernetes-sigs/cloud-provider-azure#642]
**Is your feature request related to a problem?/Why is this needed**
Much of the context usage in the project is ineffectual. For example, the [Azure DiskClient.Get uses a context](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/18cf8182aa5fa090b2cfa9180bdb3c97dd6c11a7/pkg/azureclients/diskclient/azure_diskclient.go#L92), but the context provided to the method is largely ineffectual. As you can see here, a call to AttachDisk first calls [`getContextWithCancel`](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/18cf8182aa5fa090b2cfa9180bdb3c97dd6c11a7/pkg/provider/azure_controller_common.go#L159), which simply constructs a [`context.Background` with cancelation](https://github.com/kubernetes-sigs/cloud-provider-azure/blob/18cf8182aa5fa090b2cfa9180bdb3c97dd6c11a7/pkg/provider/azure_utils.go#L86-L88). The cancellation is only called in a defer when calling `AzureDisk.Get`. The cancellation provides no benefit as a way of cancelling processing further down the tree of execution. 

**Describe the solution you'd like in detail**
Context should be used to time bound execution hierarchies, and are best when the start from the root of execution. Perhaps, having a context started at the CLI entry point where OS signals are used to establish the initial lifespan of the root context.

Further, context should be exposed upward from any network exposed task. If a network request or any other long running activity is called, the context should be required and passed in from higher up in the call hierarchy.

**Additional context**
By setting a consistent approach to passing context, the application will also be able to be instrumented with tracing easier. For example, Cluster API Provider Azure uses OpenTelemetry to create a distributed trace of the entire reconciliation loop for all resources, https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/8a44c6d6f7b394b7929c3ae5b0e79730609c6cf1/controllers/azurecluster_controller.go#L114-L120. This makes it super easy to observe what the operator is doing and provides insight when trying to diagnosis issues. In fact, in CAPZ, we are exporting the traces to Application Insights to visualize the running operator. We'd love to be able to something similar in Cloud Provider Azure.

/cc @craiglpeters

**Comment by feiskyer** [comment]
Thanks @devigned for reporting the issue and sharing the ideas to provide better observability. Would you like to contribute the OpenTelemetry support?

**Comment by devigned** [comment]
Definitely a possibility. I'm discussing this with @ritazh.

First thing I wanted to do is open the issue and start the discussion. I'd suspect an enhancement doc would be proper for such a change.

The major items that would need to be changed:
- Pipe `context.Context` from process entry point down through the call hierarchy to any function that may have i/o or an operation that should gracefully cancel.
- Add OpenTelemetry instrumentation throughout the call hierarchy
- Document the usage

wdyt?

**Comment by feiskyer** [comment]
@devigned thanks, LGTM

**Comment by feiskyer** [comment]
@devigned are you still available to work on this?

**Comment by devigned** [comment]
Yes, but it there are a couple things ahead of this. I think I'll be able to start it toward the end of the month.

/remove-lifecycle stale
